Support NVMe host management

P-Cao opened the pull request:
# Description
Add support for management of NVMe host


# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, pep8, linting, or security issues
- [x] I have performed Ansible Sanity test using --docker default
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Unit Test
Name                                                                        Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------------------------------------------------
plugins/modules/info.py                                                       346      3    116     16    96%
plugins/modules/nvme_host.py                                            146      3     42      5    96%

P-Cao committed:
Support NVMe host management

ansible-collections-svc commented:
Can one of the admins verify this patch?

Bhavneet-Sharma commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+        nvme_host_new_name=dict(type="str"),
+        max_num_paths=dict(type="str"),
+        max_num_sys_ports=dict(type="str"),
+        state=dict(required=True, type="str", choices=["present", "absent"]),
```
```suggestion
        state=dict(default='present', type="str", choices=["present", "absent"]),
```

P-Cao replied:
Done

Bhavneet-Sharma commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+                self.module.params[param] is not None
+                and len(self.module.params[param].strip()) == 0
+            ):
+                msg = f"Please provide valid {param}"
```
Remove please keyword from each error message,
```suggestion
                msg = f"Provide valid {param}"
```

Bhavneet-Sharma replied:
Applicable for all such occurrences.

P-Cao replied:
Done

Bhavneet-Sharma commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+        """
+        Create a new NVMe host based on the given parameters and retrieve the NVMe host details
+        """
+        if nvme_host_new_name:
```
As per Anupam's suggestion, We'll allow users to pass rename parameter in the playbook.
Flow will be like this: module will first do the create call, if it found any modification module should perform the modify call.

P-Cao replied:
So, if it does have both name and new_name, It would first create and then modify the name in one operation right?

Bhavneet-Sharma commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+            LOG.error(msg)
+            self.module.fail_json(msg=msg)
+
+    def remove(self, nvme_id):
```
Use proper name to identify the method.
```suggestion
    def remove_nvme(self, nvme_id):
```

P-Cao replied:
Done

Bhavneet-Sharma commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+            LOG.error(str(e))
+            self.module.fail_json(msg=str(e))
+
+    def perform_module_operation(self):
```
Instead of perform_module_operation, Please use the following design patterns  like we have used same in all new module, 

Reference here: https://github.com/dell/ansible-powerscale/blob/main/plugins/modules/alert_channel.py#L700

This pattern helps to reduce the complexity of code.

P-Cao replied:
Updated the pattern

Bhavneet-Sharma reviewed (changes requested):
Please have a look at sonar issues and apply fix.

P-Cao committed:
Refine NVMe host feature
- Refactor into chain pattern
- Modify UT accordingly
- Revise doc related problem
- Support check mode

P-Cao commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+
+DOCUMENTATION = r"""
+module: nvme_host
+version_added: '1.0.0'
```
Fixed

P-Cao commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+options:
+  max_num_paths:
+    description:
+    - Maximum number of paths per volume. Used to create/modify the NVMe host.
```
Fixed

P-Cao commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+    choices: ['present', 'absent']
+    required: true
+    type: str
+notes:
```
Check mode is supported now

sachin-apa replied:
Please refer new ansible guidelines on documneting check and diff mode.
https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-fields

P-Cao replied:
Use Attribute now

P-Cao commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+        "memoryAllocationFailure": null,
+        "versionInfo": null,
+        "sdcType": null,
+        "nqn": "nqn.2014-08.org.nvmexpress:uuid:79e90a42-47c9-a0f6-d9d3-51c47c72c7c1",
```
Updated

P-Cao commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,484 @@
+        # initialize the Ansible module
+        self.module = AnsibleModule(
+            argument_spec=self.module_params,
+            supports_check_mode=False,
```
Supported

P-Cao committed:
Use attribute instead of note for check and diff mode

trisha-dell commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,443 @@
+    description: Details of the NVMe host.
+    returned: When NVMe host exists
+    type: dict
+    contains:
```
The sample mentions many parameters. Please add all of these in the returned details.

P-Cao replied:
Added

trisha-dell commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,443 @@
+        NVMeHostCreateHandler().handle(nvme_host_obj, nvme_host_params, nvme_host_details)
+
+
+def main():
```
diff mode implementation seems to be missing.

P-Cao replied:
Supported

sachin-apa commented:
run e2e

P-Cao committed:
Support check mode and modify doc

sachin-apa commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,540 @@
+                self.result.update({"diff": {"before": nvme_host_details, "after": {}}})
+
+            if not self.module.check_mode:
+                LOG.info(msg=f"Failed to remove NVMe host {nvme_host_id}")
```
Does this log correct here ? Please check.

P-Cao committed:
Revise log and updated UT

P-Cao commented:
SonarQube report:
![image](https://github.com/user-attachments/assets/69589345-1a81-4bb4-86b2-f46bc9d665dc)
CheckMarx report:
https://osj-sio-03-prd.cec.lab.emc.com/job/Ansible-PowerFlex-CheckMarx/25/
![image](https://github.com/user-attachments/assets/e1c00429-dfb0-4cbb-bb2e-d9688e375b95)

Bhavneet-Sharma commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,540 @@
+version_added: '2.6.0'
+short_description: Manage NVMe hosts on Dell PowerFlex
+description:
+- Managing NVMe hosts on PowerFlex storage system includes getting details of NVMe hosts
```
```suggestion
- Managing NVMe hosts on PowerFlex storage system includes creating, getting details of NVMe hosts
```

P-Cao replied:
Updated

Bhavneet-Sharma commented on plugins/modules/nvme_host.py:
```diff
@@ -0,0 +1,540 @@
+short_description: Manage NVMe hosts on Dell PowerFlex
+description:
+- Managing NVMe hosts on PowerFlex storage system includes getting details of NVMe hosts
+  and renaming NVMe hosts.
```
```suggestion
  , modifying and deleting NVMe hosts.
```

Bhavneet-Sharma commented on tests/unit/plugins/module_utils/mock_nvme_host_api.py:
```diff
@@ -0,0 +1,64 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+"""
+Mock Api response for Unit tests of NVMe host module on Dell Technologies (Dell) PowerFlex
```
```suggestion
Mock Api response for Unit tests of NVMe host module on Dell Technologies PowerFlex
```

P-Cao replied:
Updated

P-Cao committed:
Revise doc

P-Cao committed:
Modify name and rename logic

P-Cao committed:
Add version check
